Adds encoding options for subjectPublicKey and privateKey

bhess opened the pull request:
Integrates encoding subjectPublicKey and privateKey according the following IETF drafts:

https://datatracker.ietf.org/doc/draft-uni-qsckeys-dilithium/00/
https://datatracker.ietf.org/doc/draft-uni-qsckeys-falcon/00/
https://datatracker.ietf.org/doc/draft-uni-qsckeys-kyber/00/
https://datatracker.ietf.org/doc/draft-uni-qsckeys-sphincsplus/00/

- Integrates encoding library from https://github.com/Quantum-Safe-Collaboration/qsc-key-rfc/tree/main/qsc-key-encoder.
The code is used as an external project with cmake `ExternalProject_Add`. Optionally, oqs-provider can be built without the dependency by using `-DUSE_ENCODING_LIB=OFF`. The library provides API to encode/decode according to the draft specification.

- The encodings are specified by setting environment variables. Example: `export OQS_ENCODING_DILITHIUM2=draft-uni-qsckeys-dilithium-00/sk-pk`. If no env variable are set, the default is 'no' encoding (the previous oqs-provider behavior).

- Supports the `-DNOPUBKEY_IN_PRIVKEY` build option (#83). 
Note: some PQC algorithms (Kyber, SPHINCS+) already contain the public key in their 'raw' private key. With `-DNOPUBKEY_IN_PRIVKEY=OFF`, this usually leads to an unnecessary duplication of the public key. The encodings according to the IETF drafts avoid this duplication even if `-DNOPUBKEY_IN_PRIVKEY=OFF`.

- Adds CI test runs (with / without encodings set, with / without NOPUBKEY_IN_PRIVKEY)

Will add config options to Wiki if merged.

Fixes #89

baentsch commented on scripts/runtests_encodings.sh:
```diff
@@ -0,0 +1,16 @@
+OQS_ENCODING_SPHINCSSHA256128FROBUST=draft-uni-qsckeys-sphincsplus-00/sk-pk \
+OQS_ENCODING_SPHINCSSHA256128SSIMPLE=draft-uni-qsckeys-sphincsplus-00/sk-pk \
+OQS_ENCODING_SPHINCSSHAKE256128FSIMPLE=draft-uni-qsckeys-sphincsplus-00/sk-pk \
+scripts/runtests.sh
```
Please add $@ here to allow passing command line parameters to `runtests.sh`

bhess replied:
Added in 58a2c116de9d25fbb6a53e673be5b7238b092eec

baentsch reviewed (changes requested):
Missing some documentation: Which env vars activate what (what are permissible env var values)?

baentsch commented:
> Will add config options to Wiki if merged.

Wiki documentation does not help people who just use a release build; also it becomes stale, so please document (at least the above) within github repo.

Also, the oqs-provider version information does not contain a hint as to what encoding is actually used in the built binary. Please add sth meaningful e.g. here: https://github.com/open-quantum-safe/oqs-provider/blob/a7e34648db275cdaaf2950d884ce9a5850e2ce52/oqsprov/oqsprov.c#L422.

bhess commented:
> Missing some documentation: Which env vars activate what (what are permissible env var values)?

Added a section to README.md in 58a2c116de9d25fbb6a53e673be5b7238b092eec.

> Also, the oqs-provider version information does not contain a hint as to what encoding is actually used in the built binary. Please add sth meaningful e.g. here:

Also added in 58a2c116de9d25fbb6a53e673be5b7238b092eec.

baentsch commented on oqsprov/oqsprov.c:
```diff
@@ -121,6 +121,50 @@ int oqs_patch_oids(void) {
+
+
+int oqs_patch_encodings(void) {
+///// OQS_TEMPLATE_FRAGMENT_ENCODING_PATCHING_START
```
Should this not be #ifdef'd (dependent on USE_ENCODING_LIB)?

bhess replied:
Right, this is unused if USE_ENCODING_LIB is not set. Updated with ifdefs.

baentsch commented on oqsprov/oqs_prov.h:
```diff
@@ -25,11 +25,20 @@
+#define ENCODING_LIB_STR ""
+#endif
+
+#define OQS_PROVIDER_VERSION_STR OQS_PROVIDER_BASE_VERSION_STR NOPUBKEY_IN_PRIVKEY_STR ENCODING_LIB_STR
```
OK - I asked for it -- but isn't this a bit "too much"? Presence of encoding support is already shown in line 469, right? Preferably we can drop this (as well as "-nopub" (would be nice if we could achieve agreement on this soon -- not duplicating the public key in the private key could be default as far as I'm concerned: What's your take, @dstebila @bhess? If I'm not mistaken, this behaviour is no default only because we wanted interoperability between oqs-openssl111 and oqs-provider -- but we may slowly want to drop that, right?)

bhess replied:
Removed the redundant string now (also -nopub).

Re. public key in private key: I imagine it can be useful to be able to get the public key directly from a PKCS#8 structure, e.g. for a signer who wants to test sign/verify without regenerating the pk. An advantage of the IETF drafts implemented in this PR is that the private keys are structured to have an OPTIONAL public key. If one choses to include the OPTIONAL part, there won't be a duplication of the public key. If one just concatenates the raw sk/pk, there will be duplication in some cases like Kyber/SPHINCS+.

baentsch reviewed (changes requested):
Thanks for these changes and feedback, @bhess. Will follow up with a clean-up PR for config options. Just one change request remaining (reducing size of base version string).

baentsch commented on README.md:
```diff
@@ -178,6 +178,12 @@ excludes all algorithms of the "Sphincs" family.
 
+### Key Encoding
+
+By setting `-DUSE_ENCODING_LIB=<ON/OFF>` at compile-time, oqs-provider can be compiled with with an an external encoding library `qsc-key-encoder`. Configuring the encodings is done via environment as described in [ALGORITHMS.md](ALGORITHMS.md).
```
Note to self: Document default value (ON).

baentsch commented on .circleci/config.yml:
```diff
@@ -37,6 +37,22 @@ jobs:
+      - run:
+          name: Build OQS-OpenSSL provider (only STD algs) with NOPUBKEY_IN_PRIVKEY
+          command: |
+             rm -rf _build && mkdir _build && cd _build && cmake -GNinja -DNOPUBKEY_IN_PRIVKEY=ON -DOPENSSL_ROOT_DIR=$(pwd)/../.local -DCMAKE_PREFIX_PATH=$(pwd)/../.local .. && ninja
```
Note to self: No CI test for `USE_ENCODING_LIB=OFF` -> Add one if we want to keep that (plain) option.

baentsch reviewed (approved):
Thanks again for the contribution!
